src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -1419,6 +1433,150 @@ DatabaseShardImp::removeFailedShard(std::shared_ptr<Shard> shard)
     setFileStats();
 }
 
+bool
+DatabaseShardImp::callForLedgerSQL(
+    LedgerIndex ledgerSeq,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+    auto shardIndex = seqToShardIndex(ledgerSeq);
+
+    if (shards_.count(shardIndex) && shards_[shardIndex].state == ShardInfo::State::finalize)
+    {
+        return shards_[shardIndex].shard->callForLedgerSQL(callback);
+    }
+
+    return false;
+}
+
+bool
+DatabaseShardImp::callForTransactionSQL(
+    LedgerIndex ledgerSeq,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+    auto shardIndex = seqToShardIndex(ledgerSeq);
+
+    if (shards_.count(shardIndex) && shards_[shardIndex].state == ShardInfo::State::finalize)
+    {
+        return shards_[shardIndex].shard->callForTransactionSQL(callback);
+    }
+
+    return false;
+}
+
+bool
+DatabaseShardImp::iterateLedgerSQLsForward(
+    LedgerIndex minShardIndex,
+    std::function<bool(soci::session& session, LedgerIndex index)> const& callback)
+{
+    std::lock_guard lock(mutex_);
+
+    std::map<std::uint32_t, ShardInfo>::iterator it, eit;
+
+    if (minShardIndex == (LedgerIndex)-1)
```

Comment:
I'd rather use an `optional` instead of `-1` to mean the first stored shard.

Reply:
Done.